internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -336,7 +354,53 @@ func (r *BMACReconciler) handleBMHFinalizer(ctx context.Context, log logrus.Fiel
 		return reconcileComplete{stop: true, dirty: true}
 	}
 
-	if dirty := r.configureBMHForCleaning(ctx, log, bmh); dirty {
+	drainTimedOut, err := nodeDrainTimedout(bmh)
+	if err != nil {
+		message := fmt.Sprintf("failed to check node drain timeout: %s", err.Error())
+		log.Error(message)
+		setAnnotation(&bmh.ObjectMeta, BMH_NODE_DRAIN_STATUS_ANNOTATION, message)
+		return reconcileComplete{stop: true, dirty: true}
+	}
+
+	if !drainTimedOut && bmh.GetAnnotations()[BMH_NODE_DRAIN_STATUS_ANNOTATION] != drainStatusSuccess {
+		dirty := false
+		log.Info("Draining agent node...")
+		if _, present := bmh.GetAnnotations()[BMH_NODE_DRAIN_START_ANNOTATION]; !present {
+			setAnnotation(&bmh.ObjectMeta, BMH_NODE_DRAIN_START_ANNOTATION, time.Now().UTC().Truncate(time.Second).Format(time.RFC3339))
+			dirty = true
+
+		}
+		requeue, err := r.drainAgentNode(ctx, log, agent)
+		if err != nil {
+			errString := fmt.Sprintf("failed to drain node: %s", err.Error())
+			if val := bmh.GetAnnotations()[BMH_NODE_DRAIN_STATUS_ANNOTATION]; val != errString {
+				setAnnotation(&bmh.ObjectMeta, BMH_NODE_DRAIN_STATUS_ANNOTATION, errString)
+				dirty = true
+			}
+			return reconcileError{err: err, dirty: dirty}
+		}
+		if requeue {
+			if val := bmh.GetAnnotations()[BMH_NODE_DRAIN_STATUS_ANNOTATION]; val != drainStatusInProgress {
+				setAnnotation(&bmh.ObjectMeta, BMH_NODE_DRAIN_STATUS_ANNOTATION, drainStatusInProgress)
+				dirty = true
+			}
+			return reconcileRequeue{requeueAfter: 20 * time.Second, dirty: dirty}
+		}
+	}
+
+	dirty := false
+	if val := bmh.GetAnnotations()[BMH_NODE_DRAIN_STATUS_ANNOTATION]; val != drainStatusTimeout && val != drainStatusSuccess {
+		if drainTimedOut {
+			log.Warn("Timed out waiting to drain node, continuing with deprovisioning")
+			setAnnotation(&bmh.ObjectMeta, BMH_NODE_DRAIN_STATUS_ANNOTATION, drainStatusTimeout)
+			dirty = true
+		} else {
```

Comment:
Why do we set the status to success?
How do we know the drain completed successfully and not still in progress?

Reply:
If the drain is still in progress `drainAgentNode` would have returned `requeue == true` and we would have returned from the function on line 387.

The only way we get here is if the drain returns successfully.
I can see if the logic can be improved here, but I remember this being a bit of a pain.

Reply:
I refactored this in the new function. Hopefully it's a bit more clear now.